syntax = "proto3";

package skylync;

message Context {
    uint64 refId = 1;
}

message AttachParams {
    string id = 1;
    string pwd = 2;
}

message OperationRequestParams {
    uint64 refId = 1;
}

message DeviceId {
   uint64 ref = 1;
   string name = 2;
}

message DeviceList {
    repeated DeviceId devices = 1;
}

message Message {
  enum Command {
    UNKNOWN_COMMAND = 0;
    ACCEPT = 1;
    REJECT = 2;
    ATTACH = 3;
    RELEASE = 4;
    CONTEXT_UPDATE = 5;
    DEVICE_LIST = 6;
    DEVICE_STATUS = 7;
    DEVICE_UPDATE_STATUS = 8;
    OPERATION_REQUEST = 9;
    OPERATION_TEARDOWN = 10;
    OPERATION_STARTED = 11;
    TUNNEL_REQUEST = 12;
    CHANNEL_OPEN = 13;
    CHANNEL_VALIDATE = 14;
    HO_REQUEST = 15;
  }

  enum Cause {
    UNKNOWN_CAUSE = 0;
    BAD_DATA = 1;
    INTERNAL_ERROR = 2;
    NOT_FOUND = 3;
    UNAUTHORIZED = 4;
    BUSY = 5;
    BAD_DEVICE = 6;
    NOT_CAPABLE = 7;
    NOTED = 8;
    BLOCKED = 9;
    ALLREADY_IN_USE = 10;
    OVERLOADED = 11;
    AUTHENTICATION_FAILED = 12;
  }

  Command command = 1;

  Command responseFor = 2;
  Cause cause = 3;

  string note = 4;
}

message EndpointMessage {
  Message base = 1;

  AttachParams attachParams = 2;
  Context context = 3;
  OperationRequestParams operationRequestParams = 4;
}

message BridgeMessage {
  Message base = 1;

  Context context = 2;
  DeviceList deviceList = 3;
}
